pkg/certwatcher/watcher.go

```diff
@@ -0,0 +1,79 @@
+package certwatcher
+
+import (
+	"context"
+	"crypto/tls"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"go.uber.org/zap"
+)
+
+type certWatcher struct {
+	cert              *tls.Certificate
+	certFile, keyFile string
+	log               *zap.SugaredLogger
+
+	mutex sync.RWMutex
+}
+
+// GetCertificate returns the certificate.
+func (w *certWatcher) GetCertificate(_ *tls.ClientHelloInfo) (*tls.Certificate, error) {
```

Comment:
shouldn't it return a copy of certs? otherwise there is no much sense to protect a pointer by mutex.

Reply:
> otherwise there is no much sense to protect a pointer by mutex

Why not? You're just synchronising the timing here, which guarantees that even the pointer read will give you a consistent result. As long as the reader reads it at the "right" time, even the pointer gives a correct result. And the "right" timing is ensured by the lock, so I don't see a problem here

Reply:
Hmm, sure, yeah.. the caller can change the pointer which bypasses the lock protection.

But I am returning a pointer here mainly to comply with the signature needed in the `TLSConfig`, which is `GetCertificate func(*ClientHelloInfo) (*Certificate, error)`.

Moreover, [this](// Once a Certificate is returned it should not be modified.) comment on `GetCertificate` says:
```
// Once a Certificate is returned it should not be modified.
```

and we are complying with it, so I don't think its a big deal in this case. But for the sake of complete correctness, I'm happy to change it

Reply:
Updated to return a pointer to a copy of the certs